src/common/stats/StatsManager.h

```diff
@@ -39,7 +73,7 @@ namespace stats {
  *                           in the last one minute
  *   error.count.600    -- Total number of errors in the last ten minutes
  */
-class StatsManager final {
+class StatsManager final : public PrometheusSerializer {
```

Comment:
If we want to add another metrics formats for `StatsManager` in the future,  how to do that ?  Inherit from another seriallizer? So I think that composition is better than inheritance for relationship between StatsManager and serializer.

Reply:
And the special `prometheusSerialize` function is not necessary to be duplicate defined. Use `serialize` directly.

Reply:
If more than one serialize format, we need change function `serialize` to such as `XXXSerialize`, and call by `serializer->XXXSerialize` to determine which function call.

Reply:
Your MetricsSerializer only has one serialize interface, how to do that like what you said? Add another XXXSerialize interface in MetricsSerializer? Does StatsManager not need to inherit a XXXSerializer?

Reply:
The MetricSerializer will contains multiple XXXSerialize function which inherited by XXXSerializer and implement by StatsManager finally.

Reply:
Agreed, better to have a member `serializer` which inherit from `PrometheusSerializer`.

Reply:
> Agreed, better to have a member `serializer` which inherit from `PrometheusSerializer`.

But the PrometheusSerializer don't know the counter and histogram.I don't think this inherit is awful. Without inherit could pass the StatsManager to `prometheusSeriaize` method.

Reply:
And I don't think the StatsManager is a serializer.

Reply:
> > The MetricSerializer will contains multiple XXXSerialize function which inherited by XXXSerializer and implement by StatsManager finally.
> 
> I don't think so. If like that to expand new serializer, the MetricSerializer abstraction layer is meaningless .

The `MetricSerializer` layer contains only interface to user, the `XXXSerializer` layer contains the utils function to serialize follow the specified format such as `writeHead` function in `PrometheusSerializer`, I don't think the two layers is extra design.

Reply:
> And I don't think the StatsManager is a serializer.

Yes, I think so. It don't follow the intuition and is not friendly to use.